SDKS-4300 Implement Device Profile Callback for AIC Journey

vibhorgoswami opened the pull request:
# JIRA Ticket

[SDKS-4300](https://pingidentity.atlassian.net/browse/SDKS-4300)
Also implements, [SDKS-3549](https://pingidentity.atlassian.net/browse/SDKS-3549)

# Description
This PR collects device profile information:
- Device Profile Callback for the AIC Journey so that device information can be collected and leveraged for device matching, while allowing flexibility for developers to provide custom attributes
- Collect the device's geographic location (longitude and latitude), so that the device profile can be enriched with real-time location data.

Does not include: Root Detection

witrisna committed:
SDKS-4300 Module setup for AIC Journey Device Profile Callback

witrisna committed:
SDKS-4300 Update README.md for the design idea

vibhorgoswami committed:
Add device profile callback and collection mechanism for hardware and platform.

vibhorgoswami committed:
Add NetworkCollector and TelephonyCollector

witrisna committed:
Refactor DeviceProfile and enhance README

This commit introduces the following changes:

- Refactored `DeviceProfile.kt` to simplify the `profile` function and improve the structure of `DeviceProfileResult`. The `metadata` is now directly a `JsonElement` and the `Json` instance is moved to a top-level property.
- Significantly enhanced `README.md` with more detailed explanations, including:
    - Features overview
    - Table of Contents
    - Permissions information
    - Detailed sections on built-in collectors with example JSON output
    - Advanced custom collector examples
    - Conditional collector loading examples
    - Improved AIC Journey Integration section with profile structure and error handling
    - Comprehensive API Reference section
- Updated `DeviceProfileCallback.kt` to return a `Result<JsonObject>` from the `collect` function, allowing for better error handling.
- Added an extension function `collect()` for `MutableList<DeviceCollector<*>>` in `DeviceCollector.kt` to simplify the collection of data from multiple collectors and return a `JsonElement`.
- Included a new test case `extension function test` in `DeviceProfileCallbackTest.kt` to verify the new collector extension function.

vibhorgoswami committed:
Fix unit test cases by introducing mocks.

vibhorgoswami committed:
Add LocationCollector for default collection.

vibhorgoswami committed:
Fix test case for location.

vibhorgoswami committed:
DeviceProfileCallback UI now requests permission for location so that we can use the LocationCollector.

vibhorgoswami committed:
Add LocationRequestActivity for collecting the location permission from the user. Also added a LocationCollectorTest

vibhorgoswami committed:
Add documentation for methods. Fix race condition with location permission activity.

vibhorgoswami committed:
Add tests for collectors.

vibhorgoswami committed:
Rebase

vibhorgoswami committed:
Fix crash with DeviceProfile collecting deviceIdentifier string.

vibhorgoswami committed:
Fix failing Unit test for DeviceProfileCallback

vibhorgoswami committed:
Fix failing Unit test for HardwareCollectorTest

vibhorgoswami committed:
Revert integration-tests.yaml

vibhorgoswami committed:
PR Review

vibhorgoswami committed:
PR Review - Add CallbackInitializer for registering Callback similar to external-idp.

vibhorgoswami commented on foundation/device/device-profile/src/main/kotlin/com/pingidentity/device/profile/collector/LocationCollector.kt:
```diff
@@ -0,0 +1,148 @@
+ * @return [LocationInfo] with coordinates if successful, null if any step fails
+ */
+@SuppressLint("MissingPermission")
+private suspend fun getLocation(): LocationInfo? {
```
Updated to use with Lock

vibhorgoswami commented on foundation/device/device-profile/src/main/kotlin/com/pingidentity/device/profile/DeviceProfile.kt:
```diff
@@ -0,0 +1,46 @@
+
+}
+
+val json = Json
```
WIP

vibhorgoswami commented on foundation/device/device-profile/src/main/kotlin/com/pingidentity/device/profile/DeviceProfileCallback.kt:
```diff
@@ -0,0 +1,59 @@
+
+        if (metadata) {
+            input(prettyString)
+        }
```
I will compare this with the FR SDK and see how to implement.

witrisna commented on foundation/device/device-profile/build.gradle.kts:
```diff
@@ -0,0 +1,38 @@
+    implementation(libs.kotlinx.serialization.json)
+    implementation(libs.androidx.annotation)
+    implementation(libs.androidx.appcompat)
+    implementation(libs.play.services.location)
```
The runtime is from the Application, so the Application needs to include the dependency if they want to capture location. The SDK set it to compileOnly just for compile the SDK.

vibhorgoswami committed:
Remove hard dependency and make it optional for developers.

vibhorgoswami committed:
Request both COARSE and FINE locations.

vibhorgoswami committed:
Fix doc for permission change.

vibhorgoswami committed:
PR review - changes for DeviceProfileCallback and DeviceProfile configuration.

vibhorgoswami committed:
Use DeviceProfileCollector and deliver result instead of using DeviceProfile

vibhorgoswami committed:
Fix DeviceProfileCallback

vibhorgoswami committed:
Fix display metrics collection for Android R and above. Add tests and documentation for DeviceProfileCollector.

vibhorgoswami committed:
Add tests for DeviceProfileCallback and document existing tests.

vibhorgoswami commented on foundation/device/device-profile/src/main/kotlin/com/pingidentity/device/profile/collector/HardwareCollector.kt:
```diff
@@ -0,0 +1,152 @@
+        val windowManager = ContextProvider.context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
+        try {
+            val metrics = DisplayMetrics()
+            windowManager.defaultDisplay.getMetrics(metrics)
```
I've updated to check version and use non-deprecated method.

vibhorgoswami committed:
Add timeout parameter to LocationCollector and getLocation method

vibhorgoswami commented on foundation/device/device-profile/src/main/kotlin/com/pingidentity/device/profile/collector/LocationCollector.kt:
```diff
@@ -0,0 +1,148 @@
+
+            // Suspend and wait for the Activity to complete the deferred.
+            // Add a timeout to prevent indefinite waiting if something goes wrong
+            val isGranted = kotlinx.coroutines.withTimeoutOrNull(30000) { // 30 second timeout
```
Updated

vibhorgoswami committed:
Rename metadata to collectors. Add BluetoothCollector and BrowserCollector. Add documentation for DeviceProfileCallback and various collectors.

witrisna reviewed (changes requested):
Add Test for Bluetooth, Browser Collectors.

vibhorgoswami committed:
Move test files to the collector package and add tests for Bluetooth and Browser collectors.

vibhorgoswami committed:
Configure Json serialization in DeviceProfileCallback.

This change configures the Json object in `DeviceProfileCallback.kt` to ignore unknown keys and not include explicit nulls during serialization.

vibhorgoswami commented on foundation/device/device-profile/src/main/kotlin/com/pingidentity/device/profile/DeviceProfileCallback.kt:
```diff
@@ -0,0 +1,358 @@
+        config.apply(block)
+
+        val result = DeviceProfileCollector(config)
+        val profile = Json.encodeToJsonElement(result.collect()).jsonObject
```
Updated to ensure that we don't allow nulls. Also added Bluetooth and Browser Collector tests.

vibhorgoswami committed:
Update README.md with BluetoothCollector and BrowserCollector.

The `DeviceProfileCallback.collect()` method now requires developers to explicitly pass the desired collectors instead of using default collectors.
The `metadata` block in the DSL for customizing device profile collection has been renamed to `collectors`.

vibhorgoswami committed:
Update README.md

vibhorgoswami committed:
Remove unused library import.

vibhorgoswami committed:
Rebased off develop.
Add test for `AndroidBuildProvider`. Update `play-services-location` and `kotlinx-coroutines-play-services` dependencies.

vibhorgoswami committed:
Revert change for integration-tests.yaml

vibhorgoswami committed:
Set DefaultDeviceCollector in DeviceProfileCallback.

witrisna commented on foundation/device/device-profile/src/main/kotlin/com/pingidentity/device/profile/DeviceProfileCallback.kt:
```diff
@@ -0,0 +1,363 @@
+     * @see DeviceProfileConfig
+     * @see DeviceProfileCollector
+     */
+    suspend fun collect(block: DeviceProfileConfig.() -> Unit = { DefaultDeviceCollector() }): Result<JsonObject> {
```
Should not include the DefaultDeviceCollector() here, it has no use.

witrisna commented on foundation/device/device-profile/src/main/kotlin/com/pingidentity/device/profile/DeviceProfileCallback.kt:
```diff
@@ -0,0 +1,363 @@
+        val config = DeviceProfileConfig()
+        config.metadata = metadata
+        config.location = location
+
```
May apply the default under com.pingidentity.device.profile.DeviceProfileConfig

```
 val collectors: MutableList<DeviceCollector<*>> = mutableListOf<DeviceCollector<*>>().apply(DefaultDeviceCollector())
```

vibhorgoswami replied:
Added the change.

vibhorgoswami committed:
Add `clear()` to DeviceProfileCallback. Add test for `WebSettings.getDefaultUserAgent()`

vibhorgoswami committed:
Add `clear()` method call in DeviceProfileCallback and update logger constants

vibhorgoswami committed:
Fix tests for collectors.

vibhorgoswami committed:
Add test cases for metadata collection.

spetrov reviewed (approved):
LGTM! 👍🏻 
- Please, let's confirm the `since` annotation usage in the docs. We haven't used this so far...
- Also do a global search for the extra space in the copyright header...

vibhorgoswami committed:
Remove @since KDoc tags and fix copyright notices.
